### PR TITLE
sew.lic: Add magic training support while crafting.

### PR DESCRIPTION
**Pull request by Sarvatt**
Uses the crafting_training_spells: yaml setting, I initially made it use training-spells with discerns but charging cambrinth isn't worth the time unless you really want to.

### CODE REVIEW

**Comment by Sarvatt** [comment]
don't review this yet please, it's a mess :)

**Comment by Sarvatt** [comment]
This is ready for review now, apologies for submitting it too early.

File: sew.lic [hunk 1]
```diff
@@ -298,9 +301,77 @@ class Sew
         tool = 'sewing needles'
       end
     end
+    magic_routine
```
[the remark this answers is not in the document]
Comment (Sarvatt) [reply sew.lic]: that is handled in the updated version

File: sew.lic [hunk 2]
```diff
@@ -51,11 +52,13 @@ class Sew
     @noun = args.noun
     @mat_type = args.material
     @stamp = @settings.mark_crafted_goods
+    @training_spells = @settings.crafting_training_spells
```
[the remark this answers is not in the document]
Comment (Sarvatt) [reply sew.lic]: base and base-empty updated now, not sure about merging data in dependency, I'm not seeing where that applies?

File: sew.lic [hunk 3]
```diff
@@ -298,9 +304,90 @@ class Sew
...
+    cast?(@data['cast'], @data['symbiosis'], @data['before'], @data['after'])
+    Flags.reset('magic-ready')
+    @preparing = false
+  end
+
+  def prepare_spell
+    return if Flags['magic-ready']
+
+    @preparing = true
+    needs_training = %w[Warding Utility Augmentation]
```
Comment (Sarvatt) [inline sew.lic]: I'm not sure if we want to allow Sorcery here. Thoughts?

Comment (dr-leustyin) [reply sew.lic]: I'm thinking no, unless you want to deal with the logic to get blown-off hands fixed as well

Comment (dr-leustyin) [reply sew.lic]: Also, there's potential for lost crafting tools if both hands go away

Comment (jayRyan24) [reply sew.lic]: The main issue is that if you're crafting in town, which is probably something many Traders are doing, it's illegal to use Sorcery. So, another compounded issue.

**Comment by jayRyan24** [comment]
Do this across the board so I can make a trader, thanks my dude.

File: sew.lic (hunk 2 above)
[the remark this answers is not in the document]
Comment (rpherbig) [reply sew.lic]: ❓ : https://github.com/rpherbig/dr-scripts/blob/master/dependency.lic#L843 has some logic to merge spell settings with `base-spells`.

You'll probably want to add a test to `validate.lic` too.

File: sew.lic [hunk 4]
```diff
@@ -367,6 +455,7 @@ class Sew
     when /trash/
       dispose_trash(@noun)
     end
+    magic_cleanup if @training_spells
```
Comment (rpherbig) [inline sew.lic]: 💡 I recommend moving `if @training_spells` into `magic_cleanup` as a guard clause

File: sew.lic (hunk 3 above)
Comment (rpherbig) [reply sew.lic]: I'm on board with not allowing Sorcery.

**Comment by rpherbig** [comment]
> I want to abstract this eventually to use in all crafting scripts and potentially other things like performance that Leus brought up wanting another spell set where you don't use cambrinth since you can cast while performing as long as you don't use it, but I wanted to start here first. I'm not sure if I'll be able to pull it out into common-arcana properly myself.

I'm fine with merging this code into `sew.lic`, but I'd really like to see it refactored into a more reusable state, if possible, before we roll it out to the other crafts.

File: sew.lic (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sarvatt) [reply sew.lic]: done and done

File: dependency.lic [hunk 5]
```diff
@@ -877,6 +877,17 @@ class SetupFiles
                                        end
     end
 
+    s.crafting_training_spells
+     .each do |skill, data|
+      s.crafting_training_spells[skill] = if spells[data['name']]
+                                         spells[data['name']].merge(data)
```
Comment (jayRyan24) [inline dependency.lic]: Is the alignment of all of this off, or is it just me?

File: validate.lic [hunk 6]
```diff
@@ -117,6 +117,13 @@ class DRYamlValidator
             .each { |skill_name| error("Invalid crafting_training_spells: skill name '#{skill_name}'. Valid skill names are '#{valid_crafting_training_spell_skills}'") }
   end
 
+  def assert_that_crafting_training_spells_have_no_cambrinth(settings)
+    return unless settings.crafting_training_spells
+
+    settings.crafting_training_spells.select { |_spell, data| data['cambrinth'] }
+            .each { |_spell, _data| warn("#{_name} in crafting_training_spells uses cambrinth which increases crafting time, you may want to remove it.") }
```
Comment (rpherbig) [inline validate.lic]: ❌  By convention, an underscore in a variable name means it is not used. Seeing both `_spell` and `_data` made me wonder what was being reported in the `.each`. I don't think `_name` is defined.

Comment (Sarvatt) [reply validate.lic]: I mean it worked, I tested it, but I don't know how to fix it properly so I'll just revert that one.

Comment (rpherbig) [reply validate.lic]: I think the test was reasonable. I suspect `_name` should have been `_data['name']`.

**Pull request merged**